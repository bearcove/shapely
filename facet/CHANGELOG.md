# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [0.1.10](https://github.com/facet-rs/facet/compare/facet-v0.1.9...facet-v0.1.10) - 2025-04-10

### Fixed

- typo in README.md

### Other

- Merge branch 'main' into from-ptr
- Update doc tests, too
- peek & poke are not reexported by the main crate
- Make tests pass on stable for now...
- Add an empty test (failing on rust stable)
- Merge branch 'main' into patch-1
- Fixed doc tests
- Merge facet-opaque, facet-spez, facet-types and facet-trait back into facet-core, to allow implementing Facet for Shape

### Other

- Removed `Facet::ARCHETYPE` in favor of new `SpezEmpty` type wrapper

## [0.1.9](https://github.com/facet-rs/facet/compare/facet-v0.1.8...facet-v0.1.9) - 2025-04-10

### Other

- middle ground
- and some spacing
- ok 30
- ok no ems
- image height
- use logo only for readmes
- logo v2

## [0.1.8](https://github.com/facet-rs/facet/compare/facet-v0.1.7...facet-v0.1.8) - 2025-04-10

### Fixed

- fix readmes

### Other

- remove spacing
- no height
- Update Readmes with logos.

## [0.1.7](https://github.com/facet-rs/facet/compare/facet-v0.1.6...facet-v0.1.7) - 2025-04-10

### Other

- doctests
- Link facet-args
- show off CLI parsing example
- shaep -> shape

## [0.1.6](https://github.com/facet-rs/facet/compare/facet-v0.1.5...facet-v0.1.6) - 2025-04-10

### Other

- Am excited
- Show it's actually runtime
- Specialization demo re: pretty
- Unify top-level README
- WIP readmes
- Fix link

## [0.1.5](https://github.com/facet-rs/facet/compare/facet-v0.1.4...facet-v0.1.5) - 2025-04-09

### Other

- upgrades
- Fix YAML link in README

## [0.1.4](https://github.com/facet-rs/facet/compare/facet-v0.1.3...facet-v0.1.4) - 2025-04-09

### Other

- DUMMY => ARCHETYPE

## [0.1.3](https://github.com/facet-rs/facet/compare/facet-v0.1.2...facet-v0.1.3) - 2025-04-08

### Other

- More nostd
- nostd facet-trait
- facet-types is now no_std friendly :)
- Less experimental now
- woo almost everything is non-exhaustive

## [0.1.2](https://github.com/facet-rs/facet/compare/facet-v0.1.1...facet-v0.1.2) - 2025-04-08

### Other

- preparing for json tuples
- Add support for sensitive fields
- not using namespace runners for now
- Fix links to other repos

## [3.1.1](https://github.com/facet-rs/facet/compare/facet-v3.1.0...facet-v3.1.1) - 2025-04-05

### Fixed

- fix errors now
- fix miri/memory problems
- fix derive probably

### Other

- Add .envrc to try removing rebuilds
- Facet is unsafe
- Fix tests etc.
- woo everything builds
- getting there
- The great split
- start fixing doc tests
- Cool, the hacking guide is in
- 29 tests passed aw yiss
- clone stuff
- clone in place => clone into
- mhmh
- mhhh getting somewhere
- fun
- bigger and bigger
- switch to btparse
- time to fix those tests
- mhmhmh
- color backtrace in tests please?
- tests are made to fail
- mhmh tests are failing huh
- maps, slowlyl
- peeking a lotta things
- more vec stuff
- okay, debug and default, it's something
- mhh we regressed
- Uhhh slices work?
- well this weirdly works?
- mhkay
- traits tests look better
- more spez is going well
- mh
- mhhhmhh
- mhhhhh it's probably the uninit thing, ngl
- I'm confused now
- uhhhh
- uhhh what
- weird
- more tests
- More tests
- rename spez to traits
- mhmhmh
- Uncomment a bunch of tests
- Unreasonably happy with that tbh
- nice nice
- Use spez-like ideas to set Debug if it's set on the type
- Well that's not really const
- yessssss
- getting somewhere maybe? but only in macros, they weren't joking.
- mhhh
- ahhhhh
- alright, will this work?
- Mhh doesn't work yet
- Well the tests do pass
- welp
- don't compare strings
- uncomment some derives
- so far so god
- new structure works
- introduce init_in_place_with_capacity
- Rname more things for more consistency
- Move tests to facet proper
- whoa hey down to 111 errors
- Innards => Def

## [3.1.0](https://github.com/facet-rs/facet/compare/facet-v3.0.0...facet-v3.1.0) - 2025-03-31

### Other

- Fill in missing docs
- Document more stuff
- Improve naming
- More notes
- Fix all markdown links
- More docs
- unfuck docs
- Mhh it derives _something_
- Tuple layouts
- more tuple support
- pre-commit hook with cargo-fmt
- More complex derives work
- Real-life derive, which fails
- Support more attribute
- more derive
- shill for namespace, closes #36
- just pre-commit, just fmt
- set up cargo-husky to run cargo fmt
- specific toolchains, reformat code

## [3.0.0](https://github.com/facet-rs/facet/compare/facet-v2.0.1...facet-v3.0.0) - 2025-03-11

### Other

- Fix doc tests in README.md
- Document how to write your deserializer a little better

## [2.0.0](https://github.com/facet-rs/facet/compare/facet-v1.0.0...facet-v2.0.0) - 2025-03-11

### Other

- Change Shape.name from static str to NameFn
- Looking good!
- clean up drop impls
- Stability notes
- link to 'free of syn' campaign
- Make derive unconditional, closes #8
- Get rid of insta, closes #10
- Ensure no syn dependency (and badge about it), closes #9
- Introduce core crate
- Get rid of debug/display, closes #4
- Start implementing transparent
