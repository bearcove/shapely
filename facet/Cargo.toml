[package]
name = "facet"
version = "0.2.0"
authors.workspace = true
edition.workspace = true
rust-version.workspace = true
license.workspace = true
repository.workspace = true
description = "Re-exports the Facet trait and derive macros, along with Shape and all related types"
keywords = ["facet", "introspection", "reflection"]
categories = ["development-tools"]

[dependencies]
facet-core = { path = "../facet-core", version = "0.5.3", default-features = false }
facet-derive = { path = "../facet-derive", version = "0.1.20" }
facet-reflect = { path = "../facet-reflect", version = "0.6.2", optional = true }

[features]
reflect = ["dep:facet-reflect"]   # Enables reflection via Peek and Poke types
testfeat = []                     # Does nothing, only used for tests
std = ["facet-core/std", "alloc"] # Uses libstd and alloc
alloc = ["facet-core/alloc"]      # Enables alloc support
default = ["std"]

[dev-dependencies]
cargo-husky = { version = "1.5.0", default-features = false, features = ["user-hooks"] }
eyre = { version = "0.6.12", default-features = false }
facet-reflect = { path = "../facet-reflect", version = "0.6.2" }
