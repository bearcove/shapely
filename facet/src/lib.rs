#![cfg_attr(not(feature = "std"), no_std)]
#![warn(missing_docs)]
#![warn(clippy::std_instead_of_core)]
#![warn(clippy::std_instead_of_alloc)]
#![doc = include_str!("../README.md")]
#![cfg_attr(docsrs, feature(doc_cfg))]
#![cfg_attr(docsrs, feature(builtin_syntax))]
#![cfg_attr(docsrs, feature(prelude_import))]
#![cfg_attr(docsrs, allow(internal_features))]

#[cfg(docsrs)]
pub mod sample_generated_code;

pub use facet_core::*;

pub use facet_derive::*;

#[cfg(feature = "reflect")]
pub use facet_reflect::*;

pub mod hacking;
