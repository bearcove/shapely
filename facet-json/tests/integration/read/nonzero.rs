use std::num::NonZero;

use eyre::Result;
use facet::Facet;
use facet_json::from_str;
use insta::assert_snapshot;

#[test]
fn json_read_nonzero_one() -> Result<()> {
    facet_testhelpers::setup();

    #[derive(Facet)]
    struct Foo {
        foo: NonZero<u64>,
    }
    let json = r#"{"foo": 1}"#;
    let s: Foo = from_str(json)?;
    assert_eq!(s.foo, { const { NonZero::new(1).unwrap() } });
    Ok(())
}

#[test]
fn json_read_nonzero_zero() -> Result<()> {
    facet_testhelpers::setup();

    #[derive(Facet, Debug)]
    struct Foo {
        foo: NonZero<u64>,
    }
    let json = r#"{"foo": 0}"#;
    let result: Result<Foo> = from_str(json).map_err(Into::into);
    assert!(result.is_err());
    #[cfg(not(miri))]
    assert_snapshot!(result.unwrap_err().to_string());
    Ok(())
}
