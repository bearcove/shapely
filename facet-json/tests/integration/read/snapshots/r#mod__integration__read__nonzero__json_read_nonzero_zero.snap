---
source: facet-json/tests/integration/read/nonzero.rs
expression: result.unwrap_err().to_string()
---
[31mError:[0m Error at [33m$.foo[39m
   [38;5;246mâ•­[0m[38;5;246mâ”€[0m[38;5;246m[[0m json:1:9 [38;5;246m][0m
   [38;5;246mâ”‚[0m
 [38;5;246m1 â”‚[0m [38;5;249m{[0m[38;5;249m"[0m[38;5;249mf[0m[38;5;249mo[0m[38;5;249mo[0m[38;5;249m"[0m[38;5;249m:[0m[38;5;249m [0m[31m0[0m[38;5;249m}[0m
 [38;5;240m  â”‚[0m         [31mâ”¬[0m  
 [38;5;240m  â”‚[0m         [31mâ•°[0m[31mâ”€[0m[31mâ”€[0m While trying to convert [32mu64[39m into transparent type [34mNonZero<u64>[39m: [31minvariant not respected[39m
[38;5;246mâ”€â”€â”€â•¯[0m
